Mathlib/RingTheory/MvPolynomial/Symmetric/FundamentalTheorem.lean

```diff
@@ -0,0 +1,522 @@
+/-
+Copyright (c) 2023 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.RingTheory.MvPolynomial.Symmetric.Defs
+import Mathlib.RingTheory.MvPolynomial.Tower
+import Mathlib.Data.Finsupp.WellFounded
+
+/-!
+# The Fundamental Theorem of Symmetric Polynomials
+
+In a polynomial ring in n variables over a commutative ring, the subalgebra of symmetric
+polynomials is freely generated by the first n elementary symmetric polynomials (excluding
+the 0th, which is simply 1). This is expressed as an isomorphism with the symmetric subalgebra
+with the polynomials in n variables, `MvPolynomial.equiv_symmetricSubalgebra`.
+
+## Proof strategy
+
+We follow the alternative proof on the Wikipedia page
+https://en.wikipedia.org/wiki/Elementary_symmetric_polynomial#Alternative_proof
+
+TODO:
+details
+declaration docstrings
+move ordered monoid instances on `Lex (σ →₀ β)`
+move maxDegree API
+-/
+
+namespace MvPolynomial
+
+open BigOperators Finset
+
+variable (σ τ R : Type*) (n m : ℕ)
+
+section CommSemiring
+
+variable [CommSemiring R]
+
+noncomputable def esymmAlgHom [Fintype σ] :
+    MvPolynomial (Fin n) R →ₐ[R] symmetricSubalgebra σ R :=
+  aeval (fun i => ⟨esymm σ R (i + 1), esymm_isSymmetric σ R _⟩)
+
+variable {σ τ R n}
+
+lemma esymmAlgHom_apply [Fintype σ] (p : MvPolynomial (Fin n) R) :
+    (esymmAlgHom σ R n p).val = aeval (fun i : Fin n => esymm σ R (i + 1)) p :=
+  (Subalgebra.mvPolynomial_aeval_coe _ _ _).symm
+
+lemma rename_esymmAlgHom [Fintype σ] [Fintype τ] (e : σ ≃ τ) :
+    (rename_symmetricSubalgebra e).toAlgHom.comp (esymmAlgHom σ R n) = esymmAlgHom τ R n := by
+  refine' algHom_ext (fun i => Subtype.ext _)
+  simp_rw [AlgHom.comp_apply, esymmAlgHom, aeval_X]
+  exact rename_esymm σ R _ e
+
+section maxDegree
+
+variable [LinearOrder σ] (p q : MvPolynomial σ R)
+
+instance (β) [CanonicallyOrderedAddMonoid β] : OrderBot (Lex (σ →₀ β)) where
+  bot := 0
+  bot_le _ := Finsupp.toLex_monotone bot_le
+
+noncomputable instance (β) [LinearOrderedCancelAddCommMonoid β] :
+    LinearOrderedCancelAddCommMonoid (Lex (σ →₀ β)) :=
+  { (inferInstance : LinearOrder (Lex (σ →₀ β))) with
+    add_le_add_left := fun a b h c => by
+      obtain rfl | ⟨i, h⟩ := h.eq_or_lt
+      · rfl
+      exact le_of_lt ⟨i, fun j hj => congr_arg (ofLex c j + ·) (h.1 j hj), add_lt_add_left h.2 _⟩
+    le_of_add_le_add_left := fun a b c h => by
+      obtain h | ⟨i, h⟩ := h.eq_or_lt
+      · exact (add_left_cancel h).le
+      exact le_of_lt ⟨i, fun j hj => (add_left_cancel <| h.1 j hj), lt_of_add_lt_add_left h.2⟩ }
```

Comment:
This probably belongs in an earlier finsupp file; preferably with identical results for `Prod`, `Pi`, and `Dfinsupp` to keep the API for `Lex` in sync.

Reply:
#6625